### ISSUE TITLE
:test_tube: InternViT, Snowflake를 활용한 T2V Projection 모델 준비

### ISSUE THREAD

**Issue by Alexaqua7** [boostcampaitech7/level4-cv-finalproject-hackathon-cv-01-lv3#61]
## 사용 방법
- t2v/에 위치한 download.py를 실행시켜, 필요한 model weights를 다운로드 해주세요

## 📌 가설 (Hypothesis)

- 가설: 사용자가 Text로 Query를 입력하면, Vision Feature와 Text Embedding을 동일한 임베딩 공간에 정렬(Align)한 Retrieval 모델(Our Model)을 통해 가장 유사한 Video Segment를 찾아낸다. 동시에, 공간 효율 향상: 원본 feature 전체 저장에 비해 용량 감소, 검색 속도 개선: 임베딩 비교만으로 빠른 유사도 계산 가능, Text-to-Video & Video-to-Text 양방향 확장 가능이 될 것이라고 생각한다.

---

## ✅ 어떻게 실험을 진행할 것 인가요?

- [x] InternVL 기반 Vision Feature 추출 완료 → N, 256, 3584
- [x] Snowflake 기반 Text Embedding 추출 완료 → 768
- [ ] Vision-Text Alignment 모델 구축 중
- [ ] VATEX 일부로 학습 데이터 구성 중

## 참조
•	[2104.00650](https://arxiv.org/pdf/2104.00650): Frozen in Time: A Joint Video and Image Encoder for End-to-End Retrieval
•	[Scaling Up Visual and Vision-Language Representation Learning With Noisy Text Supervision (https://arxiv.org/pdf/2102.05918)
---

#### 정리
- Text Embedding Model은 전혀 상관 없음 (잘 맞추어져 있음)
- 아래와 같이 Frozen-in-time의 경우, Video_data의 차원이 ([Batch_size, Embed_dim])이고, 우리 InternViT에서 extract_features를 사용한다면, ([Batch_size, Number_of_frames, Patch_dim, Embed_dim])이기 때문에 **중간에 Number_of_frames와 Patch_dim은 날려줘야 한다**
- 이를 날리기 위해서 네 가지 방법을 생각하였으며, 우선순위 역시 정하였다.
| 방법                                             | 특징 및 장점   | 단점                                | 추천도                                             | 시도 우선순위|
| `mean(dim=[1,2])`                       | 간단함            | 정보 손실 큼                    | ❌ (피해 클 수 있음)                       | 4순위             |
| `flatten(F * P)`                              | 구조 보존       | Positional Encoding 필요| ✅ (temporal-aware 모델과 함께)  | 2순위             |
| Transformer로 token aggregate | 정보 보존        | 구조 추가 필요                 | ✅✅ (가장 추천)                           | 1순위            |
| 각 프레임별 CLS만 뽑고 평균        | 요약 가능        | spatial 정보 손실             | △ (용도에 따라)                              | 3순위            |

#### 레이어 별 세부 차원 (Frozen-in-time)
type(text_data): <class 'dict'>
type(video_data): <class 'torch.Tensor'>

text_data.keys: dict_keys(['input_ids', 'attention_mask'])
text_data.keys: dict_keys(['input_ids', 'attention_mask'])

- compute_text와 compute_video를 하기 이전의 shape
video_data.shape: torch.Size([16, 4, 3, 224, 224])

compute_text 함수 내부::: text_data['input_ids'].shape: torch.Size([16, 16])
compute_text 함수 내부::: self.text_model 통과 이후 text_embeddings.shape: torch.Size([16, 768])
compute_text 함수 내부::: text projection 이후의 text_embeddings.shape: torch.Size([16, 256])
VideoTransformer 내부의 Input으로 주어진 x.shape: torch.Size([16, 4, 3, 224, 224])
VideoTransformer 내 Transformer Block을 통과하기전 x.shape: torch.Size([16, 785, 768])
VideoTransformer 내 Transformer Block을 통과한 후의 (self.norm 레이어를 통과하기 전의) x.shape: torch.Size([16, 785, 768])
VideoTransformer에서 self.norm(x)를 통과한 후의 (self.norm(x)) x.shape: torch.Size([16, 785, 768])
VideoTransformer에서의 CLS Token (self.norm(x)[:,0]) x.shape: torch.Size([16, 768])
VideoTransformer에서 pre_logits을 통과한 이후의 x.shape (self.pre_logits(x).shape): torch.Size([16, 768])
VideoTransformer 내 self.forward_features를 통과한 결과 self.forward_features(x): torch.Size([16, 768])
VideoTransformer 내부에서 head를 거치고 난 결과: self.head(self.video_features(x)): torch.Size([16, 768])

compute_video 함수 내부::: self.video_model 통과 이후 video_embeddings.shape: torch.Size([16, 768])
compute_video 함수 내부::: video projection 이후의 video_embeddings.shape: torch.Size([16, 256])

compute_text와 compute_video를 한 이후의 shape

text_embeddings.shape: torch.Size([16, 256])
video_embeddings.shape: torch.Size([16, 256])

### 정량적 결과

- 

### 정성적 결과

- 

### 결론

- 

---
